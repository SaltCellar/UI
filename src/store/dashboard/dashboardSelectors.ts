import { RootState } from '../rootReducer';
import { dashboardStateKey, getQueryForWidget } from './dashboardCommon';

export const selectDashboardState = (state: RootState) =>
  state[dashboardStateKey];

export const selectWidgets = (state: RootState) =>
  selectDashboardState(state).widgets;

export const selectWidget = (state: RootState, id: number) =>
  selectWidgets(state)[id];

export const selectCurrentWidgets = (state: RootState) =>
  selectDashboardState(state).currenWidgets;

export const selectWidgetQueries = (state: RootState, id: number) => {
  const widget = selectWidget(state, id);

  return {
    previous: getQueryForWidget(widget, -2),
    current: getQueryForWidget(widget, -1),
  };
};
