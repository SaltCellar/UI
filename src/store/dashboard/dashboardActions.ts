import { createStandardAction } from 'typesafe-actions';
import { ThunkAction } from '../common';
import { reportsActions } from '../reports';
import { DashboardTab } from './dashboardCommon';
import { selectWidget, selectWidgetQueries } from './dashboardSelectors';

export const fetchWidgetReports = (id: number): ThunkAction => {
  return (dispatch, getState) => {
    const state = getState();
    const widget = selectWidget(state, id);
    const { previous, current } = selectWidgetQueries(state, id);
    dispatch(reportsActions.fetchReport(widget.reportType, current));
    dispatch(reportsActions.fetchReport(widget.reportType, previous));
  };
};

export const setWidgetTab = createStandardAction('dashboard/widget/tab')<{
  id: number;
  tab: DashboardTab;
}>();

export const changeWidgetTab = (id: number, tab: DashboardTab): ThunkAction => {
  return dispatch => {
    dispatch(setWidgetTab({ id, tab }));
    dispatch(fetchWidgetReports(id));
  };
};
