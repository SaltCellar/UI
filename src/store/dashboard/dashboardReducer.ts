import { ActionType, getType } from 'typesafe-actions';
import { setWidgetTab } from './dashboardActions';
import { DashboardWidget } from './dashboardCommon';
import { costSummaryWidget } from './dashboardWidgets';

export type DashboardAction = ActionType<typeof setWidgetTab>;

export type DashboardState = Readonly<{
  widgets: Record<number, DashboardWidget>;
  currenWidgets: number[];
}>;

export const defaultState: DashboardState = {
  currenWidgets: [costSummaryWidget.id],
  widgets: {
    [costSummaryWidget.id]: { ...costSummaryWidget },
  },
};

export function dashboardReducer(
  state = defaultState,
  action: DashboardAction
): DashboardState {
  switch (action.type) {
    case getType(setWidgetTab):
      return {
        ...state,
        widgets: {
          ...state.widgets,
          [action.payload.id]: {
            ...state.widgets[action.payload.id],
            currentTab: action.payload.tab,
          },
        },
      };
    default:
      return state;
  }
}
