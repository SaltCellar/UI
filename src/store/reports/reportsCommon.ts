import { ReportType } from 'api/reports';

export const reportsStateKey = 'reports';

export function getReportId(type: ReportType, query: string) {
  return `${type}--${query}`;
}
