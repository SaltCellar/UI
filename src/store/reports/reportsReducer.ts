import { Report } from 'api/reports';
import { AxiosError } from 'axios';
import { ActionType, getType } from 'typesafe-actions';
import { FetchStatus } from '../common';
import {
  fetchReportFailure,
  fetchReportRequest,
  fetchReportSuccess,
} from './reportsActions';

export interface CachedReport extends Report {
  timeRequested: number;
}

export type ReportsState = Readonly<{
  byId: Map<string, CachedReport>;
  fetchStatus: Map<string, FetchStatus>;
  errors: Map<string, AxiosError>;
}>;

const defaultState: ReportsState = {
  byId: new Map(),
  fetchStatus: new Map(),
  errors: new Map(),
};

export type ReportsAction = ActionType<
  | typeof fetchReportFailure
  | typeof fetchReportRequest
  | typeof fetchReportSuccess
>;

export function reportsReducer(
  state = defaultState,
  action: ReportsAction
): ReportsState {
  switch (action.type) {
    case getType(fetchReportRequest):
      return {
        ...state,
        fetchStatus: new Map(state.fetchStatus).set(
          action.payload.reportId,
          FetchStatus.inProgress
        ),
      };

    case getType(fetchReportSuccess):
      return {
        ...state,
        fetchStatus: new Map(state.fetchStatus).set(
          action.meta.reportId,
          FetchStatus.complete
        ),
        byId: new Map(state.byId).set(action.meta.reportId, {
          ...action.payload,
          timeRequested: Date.now(),
        }),
        errors: new Map(state.errors).set(action.meta.reportId, null),
      };

    case getType(fetchReportFailure):
      return {
        ...state,
        fetchStatus: new Map(state.fetchStatus).set(
          action.meta.reportId,
          FetchStatus.complete
        ),
        errors: new Map(state.errors).set(action.meta.reportId, action.payload),
      };
    default:
      return state;
  }
}
