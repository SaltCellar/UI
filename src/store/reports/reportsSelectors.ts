import { ReportType } from 'api/reports';
import { RootState } from '../rootReducer';
import { getReportId, reportsStateKey } from './reportsCommon';

export const selectReportsState = (state: RootState) => state[reportsStateKey];

export const selectReport = (
  state: RootState,
  reportType: ReportType,
  query: string
) => selectReportsState(state).byId.get(getReportId(reportType, query));

export const selectReportFetchStatus = (
  state: RootState,
  reportType: ReportType,
  query: string
) => selectReportsState(state).fetchStatus.get(getReportId(reportType, query));

export const selectReportError = (
  state: RootState,
  reportType: ReportType,
  query: string
) => selectReportsState(state).errors.get(getReportId(reportType, query));
