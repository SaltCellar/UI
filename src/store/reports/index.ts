import * as reportsActions from './reportsActions';
import { reportsStateKey } from './reportsCommon';
import {
  CachedReport,
  ReportsAction,
  reportsReducer,
  ReportsState,
} from './reportsReducer';
import * as reportsSelectors from './reportsSelectors';

export {
  ReportsAction,
  CachedReport,
  reportsActions,
  reportsReducer,
  reportsSelectors,
  ReportsState,
  reportsStateKey,
};
