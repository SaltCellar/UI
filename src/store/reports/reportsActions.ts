import { Report, ReportType, runReport } from 'api/reports';
import { AxiosError } from 'axios';
import { ThunkAction } from 'redux-thunk';
import { createStandardAction } from 'typesafe-actions';
import { FetchStatus } from '../common';
import { RootState } from '../rootReducer';
import { getReportId } from './reportsCommon';
import { selectReport, selectReportFetchStatus } from './reportsSelectors';

const expirationMS = 30 * 60 * 1000; // 30 minutes

interface ReportActionMeta {
  reportId: string;
}

export const fetchReportRequest = createStandardAction('reports/request')<
  ReportActionMeta
>();
export const fetchReportSuccess = createStandardAction('reports/success')<
  Report,
  ReportActionMeta
>();
export const fetchReportFailure = createStandardAction('reports/failure')<
  AxiosError,
  ReportActionMeta
>();

export function fetchReport(
  reportType: ReportType,
  query: string
): ThunkAction<void, RootState, void, any> {
  return (dispatch, getState) => {
    if (!isReportExpired(getState(), reportType, query)) {
      return;
    }

    const meta: ReportActionMeta = {
      reportId: getReportId(reportType, query),
    };

    dispatch(fetchReportRequest(meta));
    runReport(reportType, query)
      .then(res => {
        dispatch(fetchReportSuccess(res.data, meta));
      })
      .catch(err => {
        dispatch(fetchReportFailure(err, meta));
      });
  };
}

function isReportExpired(
  state: RootState,
  reportType: ReportType,
  query: string
) {
  const report = selectReport(state, reportType, query);
  const fetchStatus = selectReportFetchStatus(state, reportType, query);
  if (fetchStatus === FetchStatus.inProgress) {
    return false;
  }

  if (!report) {
    return true;
  }

  const now = Date.now();
  return now > report.timeRequested + expirationMS;
}
