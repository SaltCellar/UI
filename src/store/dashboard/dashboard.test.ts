jest.mock('../reports/reportsActions');

import { createMockStoreCreator } from '../mockStore';
import { reportsActions } from '../reports';
import * as actions from './dashboardActions';
import {
  dashboardStateKey,
  DashboardTab,
  getGroupByForTab,
} from './dashboardCommon';
import { dashboardReducer } from './dashboardReducer';
import * as selectors from './dashboardSelectors';
import {
  computeWidget,
  costSummaryWidget,
  storageWidget,
} from './dashboardWidgets';

const createDashboardStore = createMockStoreCreator({
  [dashboardStateKey]: dashboardReducer,
});

const fetchReportMock = reportsActions.fetchReport as jest.Mock;

beforeEach(() => {
  fetchReportMock.mockReturnValue({ type: '@@test' });
});

test('default state', () => {
  const store = createDashboardStore();
  const state = store.getState();
  expect(selectors.selectCurrentWidgets(state)).toEqual([
    costSummaryWidget.id,
    computeWidget.id,
    storageWidget.id,
  ]);
  expect(selectors.selectWidget(state, costSummaryWidget.id)).toEqual(
    costSummaryWidget
  );
});

test('fetch widget reports', () => {
  const store = createDashboardStore();
  store.dispatch(actions.fetchWidgetReports(costSummaryWidget.id));
  expect(fetchReportMock.mock.calls).toMatchSnapshot();
});

test('changeWidgetTab', () => {
  const store = createDashboardStore();
  store.dispatch(
    actions.changeWidgetTab(costSummaryWidget.id, DashboardTab.regions)
  );
  const widget = selectors.selectWidget(store.getState(), costSummaryWidget.id);
  expect(widget.currentTab).toBe(DashboardTab.regions);
  expect(fetchReportMock).toHaveBeenCalledTimes(2);
});

describe('getGroupByForTab', () => {
  test('services tab', () => {
    expect(getGroupByForTab(DashboardTab.services)).toMatchSnapshot();
  });

  test('accounts tab', () => {
    expect(getGroupByForTab(DashboardTab.accounts)).toMatchSnapshot();
  });

  test('regions tab', () => {
    expect(getGroupByForTab(DashboardTab.regions)).toMatchSnapshot();
  });

  test('unknown tab', () => {
    expect(getGroupByForTab('unknown' as any)).toMatchSnapshot();
  });
});
