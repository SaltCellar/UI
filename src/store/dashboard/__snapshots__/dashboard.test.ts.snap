// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetch widget reports 1`] = `
Array [
  Array [
    "cost",
    "filter[time_scope_units]=month&filter[time_scope_value]=-1&filter[resolution]=daily&filter[limit]=5&group_by[service]=*",
  ],
  Array [
    "cost",
    "filter[time_scope_units]=month&filter[time_scope_value]=-2&filter[resolution]=daily&filter[limit]=5&group_by[service]=*",
  ],
]
`;

exports[`getGroupByForTab accounts tab 1`] = `
Object {
  "account": "*",
}
`;

exports[`getGroupByForTab regions tab 1`] = `
Object {
  "region": "*",
}
`;

exports[`getGroupByForTab services tab 1`] = `
Object {
  "service": "*",
}
`;

exports[`getGroupByForTab unknown tab 1`] = `Object {}`;
